functionary/builder/utils.py

```diff
@@ -151,26 +187,32 @@ def build_package(build_id: UUID):
         _extract_package_contents(package_contents, workdir)
         _load_dockerfile_template(dockerfile, workdir)
 
-        image, build_log = docker_client.images.build(
-            path=workdir,
-            pull=True,
-            forcerm=True,
-            tag=full_image_name,
+        image, build_result = docker_client.images.build(
+            path=workdir, pull=True, forcerm=True, tag=full_image_name
         )
-        docker_client.images.push(full_image_name)
-    except Exception:
+        push_result = docker_client.images.push(full_image_name, stream=True)
+        build.status = Build.COMPLETE
+    except BuildError as exc:
+        error_info = [str(line) for line in exc.build_log]
+        build_log_message = "".join(error_info)
+        build.status = Build.ERROR
+    except APIError as exc:
+        build_log_message = str(exc)
         build.status = Build.ERROR
-        build.save()
-        return
 
     with transaction.atomic():
         # Build has succeeded, save all the things now
-        for func in db_functions:
-            func.save()
-        package.image_name = image_name
-        package.save()
-        build.status = Build.COMPLETE
+        if build.status == Build.COMPLETE:
+            for func in db_functions:
+                func.save()
+            package.image_name = image_name
+            package.save()
+
+            build_log_message = _format_log_string(build_result, push_result)
+
+        log = BuildLog.objects.create(build=build, log=build_log_message)
         build.save()
+        log.save()
```

Comment:
You don't need this save, since the `BuildLog.objects.create` call does the save for you.  And since there's no use of `log` besides this, you can drop that assignment as well.